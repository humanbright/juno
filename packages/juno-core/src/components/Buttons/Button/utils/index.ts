export * from './ButtonUtils';
export * from './getButtonIconSize';
