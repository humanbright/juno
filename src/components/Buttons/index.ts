export * from './Button';
export * from './ButtonBar';
export * from './ButtonBase';
export * from './ButtonGroup';
export * from './FabButton';
export * from './FabIconButton';
export * from './IconButton';
export * from './IconButtonGroup';
export * from './RoundButton';
export * from './SplitButton';
export * from './ToggleButton';
export * from './ToggleButtonGroup';
