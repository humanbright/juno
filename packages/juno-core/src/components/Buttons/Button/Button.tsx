import React, {
  ComponentProps,
  forwardRef,
  useCallback,
  useLayoutEffect,
  useMemo,
  useRef,
} from 'react';

import MuiButton from '@material-ui/core/Button';

import {
  combineClasses,
  isRcElement,
  RcBaseColor,
  RcBaseProps,
  RcPaletteProp,
  RcTheme,
  removeClassName,
  styled,
  useDeprecatedCheck,
  useForkRef,
  useTheme,
  useThemeProps,
  RcBaseFocusVariant,
} from '../../../foundation';
import { RcIcon, RcIconProps } from '../../Icon';
import { RcIconSizes } from '../../Icon/utils';
import { RcCircularProgress, RcCircularProgressProps } from '../../Progress';
import { withTooltip } from '../../Tooltip';
import { buttonStyle } from './styles';
import type { RcButtonSize } from './utils';
import { getButtonIconSize, RcButtonClasses } from './utils';

type RcButtonColor = RcBaseColor<
  'primary' | 'secondary' | 'negative' | 'positive' | 'action' | 'neutral'
>;

type RcButtonColorLoadingMode = 'prefix' | 'replace' | 'suffix';

type RcButtonVariant = 'text' | 'outlined' | 'contained' | 'plain';

type RcButtonRadius = keyof RcTheme['radius'];

type RcButtonProps = {
  /** size of button, default is `large` */
  size?: RcButtonSize;
  /** variant of button, default is `contained` */
  variant?: RcButtonVariant;
  /** color of button */
  color?: RcPaletteProp | RcButtonColor;
  /** is button loading, when `loading` that `disabled` will auto become `true` */
  loading?: boolean;
  /**
   * disabled variant
   *
   * - `normal`: default disabled button style
   * - `mask`: with mask cover on button, and do not change the button base color
   *
   * @default 'normal'
   */
  disabledVariant?: 'normal' | 'mask';
  /** loading mode with button, default is `replace` */
  loadingMode?: RcButtonColorLoadingMode;
  /** Props send to `RcCircularProgress` when loading is `true` */
  CircularProgressProps?: RcCircularProgressProps;
  /** component for render root button, default is `button` */
  component?: React.ElementType;
  /** border radius for button */
  radius?: RcButtonRadius;
  /** should keep elevation when type is `container` */
  keepElevation?: boolean;
  /** @deprecated Icon, please use startIcon with `<RcIcon />` */
  IconProps?: RcIconProps;
  /**
   * Set focus style for component. Note: plain variant always use focusRing
   *
   * @default 'ripple'
   */
  focusVariant?: RcBaseFocusVariant<'focusRing' | 'ripple'>;
} & RcBaseProps<
  ComponentProps<typeof MuiButton>,
  | 'color'
  | 'size'
  | 'variant'
  | 'title'
  // * use disableFocusRipple, so omit that
  | 'focusRipple'
>;

const _RcButton = forwardRef<any, RcButtonProps>(
  (props: RcButtonProps, ref) => {
    if (process.env.NODE_ENV !== 'production') {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useDeprecatedCheck(RcButton, props, [
        {
          prop: 'IconProps',
          time: '2021-4',
          comment: `@deprecated Icon, please use startIcon with \`<RcIcon />\` */`,
        },
      ]);
    }
    const {
      children: childrenProp,
      classes: classesProp,
      startIcon: startIconProp,
      endIcon: endIconProp,
      loading,
      loadingMode,
      disabled,
      IconProps,
      size,
      color,
      CircularProgressProps,
      variant,
      disabledVariant,
      radius,
      keepElevation,
      focusVariant,
      disableFocusRipple: disableFocusRippleProp,
      ...rest
    } = props;

    const theme = useTheme();
    const innerRef = useRef<HTMLButtonElement>(null);
    const buttonRef = useForkRef(innerRef, ref);

    const isPlain = variant === 'plain';
    const isReplace = loadingMode === 'replace';
    const iconSize = getButtonIconSize(size);

    const loadingElm = useMemo(() => {
      if (loading) {
        const progressSize = RcIconSizes[iconSize];

        return (
          <RcCircularProgress
            size={progressSize}
            color="inherit"
            {...CircularProgressProps}
          />
        );
      }
      return undefined;
    }, [CircularProgressProps, iconSize, loading]);

    const getCurrIcon = useCallback(
      (currIcon: React.ReactNode) => {
        if (
          isRcElement<RcIconProps>(currIcon, ['RcIcon']) &&
          currIcon.props.size === RcIcon.defaultProps!.size
        ) {
          return React.cloneElement(currIcon, { size: iconSize });
        }

        return currIcon;
      },
      [iconSize],
    );

    const startIcon = useMemo(() => {
      if (loadingElm) {
        switch (loadingMode) {
          case 'prefix':
            return loadingElm;
          case 'replace':
            return undefined;
          default:
            break;
        }
      }

      if (startIconProp) {
        return getCurrIcon(startIconProp);
      }

      // TODO: IconProps will be remove
      if (IconProps) {
        return <RcIcon size={iconSize} {...IconProps} />;
      }

      return undefined;
    }, [
      IconProps,
      getCurrIcon,
      iconSize,
      loadingElm,
      loadingMode,
      startIconProp,
    ]);

    const endIcon = useMemo(() => {
      if (loadingElm) {
        switch (loadingMode) {
          case 'suffix':
            return loadingElm;
          case 'replace':
            return undefined;
          default:
            break;
        }
      }

      if (endIconProp) {
        return getCurrIcon(endIconProp);
      }

      return undefined;
    }, [loadingElm, endIconProp, loadingMode, getCurrIcon]);

    const classes = useMemo(
      () => combineClasses(RcButtonClasses, classesProp),
      [classesProp],
    );

    // * need remove not need MuiButton-iconSizeMedium for icon render correct
    useLayoutEffect(() => {
      removeClassName(innerRef, 'MuiButton-iconSizeMedium');
    });

    const disableFocusRipple =
      disableFocusRippleProp ?? focusVariant === 'focusRing';

    return (
      <MuiButton
        ref={buttonRef}
        disabled={disabled || loading}
        variant={isPlain ? undefined : (variant as any)}
        disableRipple={theme?.props?.MuiButton?.disableRipple || isPlain}
        startIcon={startIcon}
        endIcon={endIcon}
        classes={classes}
        disableFocusRipple={disableFocusRipple}
        {...rest}
      >
        {loading && isReplace ? loadingElm : childrenProp}
      </MuiButton>
    );
  },
);

export const RcButtonDefaultSize = 'large';
export const RcButtonDefaultColor = 'primary';
export const RcButtonDefaultVariant = 'contained';

/** @release */
const RcButton = styled(withTooltip(_RcButton)).attrs(
  (inProps: RcButtonProps) => {
    const {
      // @ts-ignore
      theme,
      // Omit className, prevent duplicate className
      className,
      size = RcButtonDefaultSize,
      color = RcButtonDefaultColor,
      variant = RcButtonDefaultVariant,
      loadingMode = 'replace',
      focusVariant = 'ripple',
      ...rest
    } = useThemeProps({ props: inProps, name: 'RcButton' });
    return {
      size,
      color,
      variant,
      loadingMode,
      focusVariant,
      ...rest,
    };
  },
)`
  ${buttonStyle}
`;

RcButton.displayName = 'RcButton';

export { RcButton };

export type {
  RcButtonColor,
  RcButtonColorLoadingMode,
  RcButtonProps,
  RcButtonSize,
  RcButtonVariant,
};
