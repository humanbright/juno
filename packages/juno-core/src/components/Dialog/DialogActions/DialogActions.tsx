import React, { ComponentProps, forwardRef, useMemo } from 'react';

import MuiDialogActions from '@material-ui/core/DialogActions';

import {
  combineClasses,
  RcBaseDirection,
  RcBaseProps,
  styled,
  useThemeProps,
} from '../../../foundation';
import { RcDialogChildrenProps, useDialogDefaultProps } from '../utils';
import { DialogActionsStyle } from './styles';
import { RcDialogActionsClasses } from './utils';

type RcDialogActionsProps = {
  /**
   * direction of actions
   *
   * @default 'horizontal'
   */
  direction?: RcBaseDirection;
  /**
   * reverse below items
   *
   * @default false, when direction is `vertical`, default reverse will be `true`
   */
  reverse?: boolean;
} & RcDialogChildrenProps &
  RcBaseProps<ComponentProps<typeof MuiDialogActions>>;

const _RcDialogActions = forwardRef<any, RcDialogActionsProps>(
  (inProps, ref) => {
    const props = useThemeProps({ props: inProps, name: 'RcDialogActions' });
    const { classes: classesProp, size, children, reverse, ...rest } = props;

    const classes = useMemo(
      () => combineClasses(RcDialogActionsClasses, classesProp),
      [classesProp],
    );

    return (
      <MuiDialogActions {...rest} ref={ref} classes={classes} disableSpacing>
        {children}
      </MuiDialogActions>
    );
  },
);

const RcDialogActions = styled(_RcDialogActions).attrs(
  (props: RcDialogActionsProps) => {
    const toProps = useDialogDefaultProps(props);

    const direction =
      toProps.direction ??
      (toProps.size === 'small' ? 'vertical' : 'horizontal');

    const reverse = toProps.reverse ?? direction === 'vertical';

    return {
      ...toProps,
      direction,
      reverse,
    };
  },
)`
  ${DialogActionsStyle}
`;

RcDialogActions.defaultProps = {};

RcDialogActions.displayName = 'RcDialogActions';

export { RcDialogActions };
export type { RcDialogActionsProps };
