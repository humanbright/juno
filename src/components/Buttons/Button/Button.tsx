import MuiButton from '@material-ui/core/Button';
import React, {
  ComponentProps,
  forwardRef,
  useCallback,
  useLayoutEffect,
  useMemo,
  useRef,
} from 'react';

import {
  combineClasses,
  isRcElement,
  RcBaseColor,
  RcBaseProps,
  RcPaletteKeys,
  RcTheme,
  removeClassName,
  styled,
  useForkRef,
  useTheme,
  useThemeProps,
  withDeprecatedCheck,
} from '../../../foundation';
import { RcIcon, RcIconProps } from '../../Icon';
import { RcIconSizes } from '../../Icon/utils';
import { RcCircularProgress, RcCircularProgressProps } from '../../Progress';
import { withTooltip } from '../../Tooltip';
import { buttonStyle } from './styles';
import { getButtonIconSize, RcButtonClasses, RcButtonSize } from './utils';

type RcButtonColor = RcBaseColor<
  'primary' | 'secondary' | 'negative' | 'positive' | 'action' | 'neutral'
>;

type RcButtonColorLoadingMode = 'prefix' | 'replace' | 'suffix';

type RcButtonVariant = 'text' | 'outlined' | 'contained' | 'plain';

type RcButtonRadius = keyof RcTheme['radius'];

type RcButtonProps = {
  /** size of button, default is `large` */
  size?: RcButtonSize;
  /** variant of button, default is `contained` */
  variant?: RcButtonVariant;
  /** color of button */
  color?: RcPaletteKeys | RcButtonColor;
  /** is button loading, when `loading` that default `disabled` will be true  */
  loading?: boolean;
  /** loading mode with button, default is `replace` */
  loadingMode?: RcButtonColorLoadingMode;
  /** Props send to `RcCircularProgress` when loading is `true` */
  CircularProgressProps?: RcCircularProgressProps;
  /** component for render root button, default is `button` */
  component?: React.ElementType;
  /** border radius for button */
  radius?: RcButtonRadius;
  /** should keep elevation when type is `container` */
  keepElevation?: boolean;
  /** @deprecated Icon, please use startIcon with `<RcIcon />` */
  IconProps?: RcIconProps;
} & RcBaseProps<
  ComponentProps<typeof MuiButton>,
  | 'color'
  | 'size'
  | 'variant'
  | 'title'
  // * use disableFocusRipple, so omit that
  | 'focusRipple'
>;

const _RcButton = forwardRef<any, RcButtonProps>(
  (inProps: RcButtonProps, ref) => {
    const props = useThemeProps({ props: inProps, name: 'RcButton' });
    const {
      children: childrenProp,
      classes: classesProp,
      startIcon: startIconProp,
      endIcon: endIconProp,
      loading,
      loadingMode,
      disabled = loading,
      IconProps,
      size,
      color,
      CircularProgressProps,
      variant,
      radius,
      keepElevation,
      ...rest
    } = props;

    const theme = useTheme();
    const innerRef = useRef<HTMLButtonElement>(null);
    const buttonRef = useForkRef(innerRef, ref);

    const isPlain = variant === 'plain';
    const isReplace = loadingMode === 'replace';
    const iconSize = getButtonIconSize(size);

    const loadingElm = useMemo(() => {
      if (loading) {
        const progressSize = RcIconSizes[iconSize];

        return (
          <RcCircularProgress
            size={progressSize}
            color="inherit"
            {...CircularProgressProps}
          />
        );
      }
      return undefined;
    }, [CircularProgressProps, iconSize, loading]);

    const getCurrIcon = useCallback(
      (currIcon: React.ReactNode) => {
        if (
          isRcElement<RcIconProps>(currIcon, ['RcIcon']) &&
          currIcon.props.size === RcIcon.defaultProps!.size
        ) {
          return React.cloneElement(currIcon, { size: iconSize });
        }

        return currIcon;
      },
      [iconSize],
    );

    const startIcon = useMemo(() => {
      if (loadingElm) {
        switch (loadingMode) {
          case 'prefix':
            return loadingElm;
          case 'replace':
            return undefined;
          default:
            break;
        }
      }

      if (startIconProp) {
        return getCurrIcon(startIconProp);
      }

      // TODO: IconProps will be remove
      if (IconProps) {
        return <RcIcon size={iconSize} {...IconProps} />;
      }

      return undefined;
    }, [
      IconProps,
      getCurrIcon,
      iconSize,
      loadingElm,
      loadingMode,
      startIconProp,
    ]);

    const endIcon = useMemo(() => {
      if (loadingElm) {
        switch (loadingMode) {
          case 'suffix':
            return loadingElm;
          case 'replace':
            return undefined;
          default:
            break;
        }
      }

      if (endIconProp) {
        return getCurrIcon(endIconProp);
      }

      return undefined;
    }, [loadingElm, endIconProp, loadingMode, getCurrIcon]);

    const classes = useMemo(
      () => combineClasses(RcButtonClasses, classesProp),
      [classesProp],
    );

    // * need remove not need MuiButton-iconSizeMedium for icon render correct
    useLayoutEffect(() => {
      removeClassName(innerRef, 'MuiButton-iconSizeMedium');
    });

    return (
      <MuiButton
        ref={buttonRef}
        disabled={disabled}
        variant={isPlain ? undefined : (variant as any)}
        disableRipple={theme?.props?.MuiButton?.disableRipple || isPlain}
        startIcon={startIcon}
        endIcon={endIcon}
        classes={classes}
        {...rest}
      >
        {loading && isReplace ? loadingElm : childrenProp}
      </MuiButton>
    );
  },
);

/** @release */
const RcButton = styled(
  withDeprecatedCheck(
    withTooltip(_RcButton),
    [
      {
        prop: 'IconProps',
        time: '2021-4',
        comment: `@deprecated Icon, please use startIcon with \`<RcIcon />\` */`,
      },
    ],
    'RcButton',
  ),
)`
  ${buttonStyle}
`;

RcButton.defaultProps = {
  size: 'large',
  color: 'primary',
  variant: 'contained',
  loadingMode: 'replace',
};

RcButton.displayName = 'RcButton';

export {
  RcButton,
  RcButtonSize,
  RcButtonProps,
  RcButtonColor,
  RcButtonColorLoadingMode,
  RcButtonVariant,
};
