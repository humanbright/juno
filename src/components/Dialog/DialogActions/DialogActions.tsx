import MuiDialogActions from '@material-ui/core/DialogActions';
import React, { ComponentProps, forwardRef, useMemo } from 'react';

import {
  combineClasses,
  RcBaseDirection,
  RcBaseProps,
  styled,
  useThemeProps,
} from '../../../foundation';
import { RcDialogProps } from '../Dialog';
import { DialogActionsStyle } from './styles';
import { RcDialogActionsClasses } from './utils';

type RcDialogActionsProps = {
  /** direction of actions */
  direction?: RcBaseDirection;
} & Pick<RcDialogProps, 'size'> &
  RcBaseProps<ComponentProps<typeof MuiDialogActions>>;

const _RcDialogActions = forwardRef<any, RcDialogActionsProps>(
  (inProps, ref) => {
    const props = useThemeProps({ props: inProps, name: 'RcDialogActions' });
    const { classes: classesProp, size, children, ...rest } = props;

    const classes = useMemo(
      () => combineClasses(RcDialogActionsClasses, classesProp),
      [classesProp],
    );

    return (
      <MuiDialogActions {...rest} ref={ref} classes={classes} disableSpacing>
        {children}
      </MuiDialogActions>
    );
  },
);

const RcDialogActions = styled(_RcDialogActions)`
  ${DialogActionsStyle}
`;

RcDialogActions.defaultProps = {
  size: 'medium',
};

RcDialogActions.displayName = 'RcDialogActions';

export { RcDialogActions, RcDialogActionsProps };
