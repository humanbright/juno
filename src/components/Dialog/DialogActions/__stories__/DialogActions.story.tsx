import { Meta, Story } from '@storybook/react';
import React, { ComponentProps } from 'react';

import {
  notControlInDocTable,
  notShowInDocTable,
  sortInDocTable,
  switchToControlKnobs,
} from '../../../../storybook';
import { RcButton } from '../../../Buttons/Button';
import { RcDialogActions } from '../DialogActions';

export default {
  title: 'ðŸš€ Cleanup Components/Dialog/DialogActions',
  component: RcDialogActions,
  argTypes: {
    ...sortInDocTable<keyof DialogActionsProps>([]),
    ...notControlInDocTable<keyof DialogActionsProps>([]),
    ...notShowInDocTable<keyof DialogActionsProps>([]),
  },
} as Meta;

type DialogActionsProps = ComponentProps<typeof RcDialogActions>;

export const DialogActions: Story<DialogActionsProps> = ({
  children,
  ...args
}) => {
  switchToControlKnobs();
  return (
    <RcDialogActions {...args}>
      <RcButton variant="text">Cancel</RcButton>
      <RcButton>Ok</RcButton>
    </RcDialogActions>
  );
};

DialogActions.storyName = 'DialogActions';

DialogActions.args = {};

DialogActions.argTypes = {
  ...notControlInDocTable<keyof DialogActionsProps>([]),
};

DialogActions.parameters = {
  tags: [
    {
      name: 'Accessibility',
      value: '100%',
    },
  ],
};
