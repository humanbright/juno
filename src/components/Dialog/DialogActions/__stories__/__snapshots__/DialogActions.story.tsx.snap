// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ðŸš€ Cleanup Components/Dialog/DialogActions DialogActions 1`] = `
.c1 {
  text-transform: none;
  font-size: 0.9375rem;
  font-weight: 700;
  font-family: Lato,Helvetica,Arial,sans-serif;
  text-align: center;
  box-shadow: unset;
  min-width: 96px;
  min-height: 36px;
  padding: 0px 16px;
}

.c1 .RcButton-startIcon {
  margin-right: 8px;
}

.c1 .RcButton-endIcon {
  margin-left: 8px;
}

.c1.RcButton-text {
  color: #066FAC;
}

.c1.RcButton-contained {
  color: #fff;
  background-color: #066FAC;
}

.c1.RcButton-contained.RcButton-disabled {
  background-color: #B9B9B9;
  color: #FFFFFF;
}

.c1.RcButton-outlined {
  color: #066FAC;
  border-color: #066FAC;
  padding: 0px 15px;
}

.c0 {
  padding: 8px 24px 24px;
}

.c0 > * + * {
  margin-left: 8px;
}

@media (hover:hover) and (pointer:fine) {
  .c1.RcButton-text:hover {
    background-color: rgba(6,111,172,0.08);
  }
}

@media (hover:hover) and (pointer:fine) {
  .c1.RcButton-contained:hover {
    background-color: rgba(6,111,172,0.92);
  }
}

@media (hover:hover) and (pointer:fine) {
  .c1.RcButton-outlined:hover {
    background-color: rgba(6,111,172,0.08);
  }
}

<div
  className="MuiDialogActions-root c0"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text RcButton-text c1"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Cancel
    </span>
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained RcButton-contained c1"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Ok
    </span>
  </button>
</div>
`;
